src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeSymbol.java

```diff
@@ -0,0 +1,60 @@
+package jdk.incubator.foreign;
+
+import jdk.internal.foreign.NativeSymbolImpl;
+import jdk.internal.reflect.CallerSensitive;
+import jdk.internal.reflect.Reflection;
+
+import java.lang.invoke.MethodHandle;
+
+/**
+ * A native symbol models a reference to a location (typically the entry point of a function) in a native library.
+ * A native symbol has a name, and is associated with a scope, which governs the native symbol's lifecycle.
+ * This is useful, since the library a native symbol refers to can be <em>unloaded</em>, thus invalidating the native symbol.
+ * While native symbols are typically obtained using a {@link SymbolLookup#lookup(String) symbol lookup}, it is also possible to obtain an
+ * <em>anonymous</em> native symbol, in the form of an {@linkplain CLinker#upcallStub(MethodHandle, FunctionDescriptor, ResourceScope) upcall stub},
+ * that is, a reference to a dynamically-generated native symbol which can be used to call back into Java code.
+ */
+sealed public interface NativeSymbol extends Addressable permits NativeSymbolImpl {
+
+    /**
+     * Returns the name of this symbol.
+     * @return the name of this symbol.
+     */
+    String name();
+
+    /**
+     * Returns the resource scope associated with this symbol.
+     * @return the resource scope associated with this symbol.
+     */
+    ResourceScope scope();
+
+    /**
+     * Returns the memory address associated with this symbol.
+     * @throws IllegalStateException if the scope associated with this symbol has been closed, or if access occurs from
+     * a thread other than the thread owning that scope.
+     * @return The memory address associated with this symbol.
+     */
+    @Override
+    MemoryAddress address();
+
+    /**
+     * Creates a new symbol from given name, address and scope.
+     * <p>
+     * This method is <a href="package-summary.html#restricted"><em>restricted</em></a>.
+     * Restricted methods are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
```

Comment:
Why is this restricted?

Reply:
because you can create a native symbol out of an arbitrary address (e.g. MemoryAddress.ofLong), which then, if passed to a downcall method handle, will cause havocs. That said, true that, at the moment you can pass Addressable to a downcall method handle - but maybe that's the issue? E.g. downcall method handle needs to take in a NativeSymbol? That way, all native symbols created by our APIs are known to be safe (upcall and symbol lookup) - but the user is still free to create custom symbols - although these latter would be restricted. Makes sense?

Reply:
That makes sense, but we are already covered there I think because creating downcall handles is a restricted operation.

Compare it to `MemoryAddress.ofLong` which is also non-restricted: it's fine to make non-dereferencable addresses, because in order to dereference them, we need to go through a restricted API. This case seems kind of similar.